File: app/schemas/procedure.py
from datetime import date as DateType
from typing import Optional, List

from pydantic import BaseModel, Field, field_validator, model_validator

from app.schemas.base_tags import TaggedEntityMixin
from app.models.enums import ProcedureStatus


class ProcedureBase(TaggedEntityMixin):
    procedure_name: str = Field(
        ..., min_length=2, max_length=300, description="Name of the procedure"
    )
    procedure_type: Optional[str] = Field(
        None,
        max_length=50,
        description="Type of procedure (e.g., surgical, diagnostic)",
    )
    procedure_code: Optional[str] = Field(
        None, max_length=50, description="Code for the procedure (e.g., CPT code)"
    )
    description: Optional[str] = Field(
        None, max_length=1000, description="Detailed description of the procedure"
    )
    date: DateType = Field(..., description="Date when the procedure was performed")
    notes: Optional[str] = Field(None, max_length=1000, description="Additional notes")
    status: str = Field(..., description="Status of the procedure")
    facility: Optional[str] = Field(
        None, max_length=300, description="Facility where the procedure was performed"
    )
    procedure_setting: Optional[str] = Field(
        None,
        max_length=100,
        description="Setting of procedure (outpatient, inpatient, office)",
    )
    procedure_complications: Optional[str] = Field(
        None,
        max_length=500,
        description="Any complications that occurred during the procedure",
    )
    procedure_duration: Optional[int] = Field(
        None, gt=0, description="Duration of the procedure in minutes"
    )
    patient_id: int = Field(..., gt=0, description="ID of the patient")
    practitioner_id: Optional[int] = Field(
        None, gt=0, description="ID of the performing practitioner"
    )
    condition_id: Optional[int] = Field(
        None, gt=0, description="ID of the condition this procedure addresses"
    )
    anesthesia_type: Optional[str] = Field(
        None, max_length=100, description="Type of Anethesia used during the procedure"
    )
    anesthesia_notes: Optional[str] = Field(
        None, max_length=1000, description="Additional notes about the anesthesia"
    )

    @model_validator(mode="before")
    @classmethod
    def validate_date_with_status(cls, values):
        """Validate procedure date based on status.

        When status is not provided (partial updates), skip validation to allow
        updating only the date field without requiring status.
        """
        from datetime import timedelta

        if not isinstance(values, dict):
            return values

        date_value = values.get("date")
        status = values.get("status", "").lower() if values.get("status") else ""

        if not date_value:
            return values

        # Convert string date to date object if needed
        if isinstance(date_value, str):
            try:
                date_value = DateType.fromisoformat(date_value)
            except ValueError:
                return values  # Let field validator handle invalid date

        # Skip validation if status is not provided (partial update scenario)
        if not status:
            return values

        # For scheduled or postponed procedures, allow reasonable future dates
        if status in ["scheduled", "postponed"]:
            max_future = DateType.today() + timedelta(days=3650)  # 10 years
            if date_value > max_future:
                raise ValueError("Procedure date cannot be more than 10 years in the future")
            # Allow past dates for scheduled procedures (e.g., rescheduled from past)
            return values

        # For all other statuses (completed, in_progress, cancelled), date should not be in future
        if date_value > DateType.today():
            raise ValueError(f"Procedure date cannot be in the future for {status} procedures")
        return values

    @field_validator("status")
    @classmethod
    def validate_status(cls, v):
        valid_statuses = [s.value for s in ProcedureStatus]
        if v.lower() not in valid_statuses:
            raise ValueError(f"Status must be one of: {', '.join(valid_statuses)}")
        return v.lower()


class ProcedureCreate(ProcedureBase):
    pass


class ProcedureUpdate(BaseModel):
    procedure_name: Optional[str] = Field(None, min_length=2, max_length=300)
    procedure_type: Optional[str] = Field(None, max_length=50)
    procedure_code: Optional[str] = Field(None, max_length=50)
    description: Optional[str] = Field(None, max_length=1000)
    date: Optional[DateType] = None
    notes: Optional[str] = Field(None, max_length=1000)
    status: Optional[str] = None
    facility: Optional[str] = Field(None, max_length=300)
    procedure_setting: Optional[str] = Field(None, max_length=100)
    procedure_complications: Optional[str] = Field(None, max_length=500)
    procedure_duration: Optional[int] = Field(None, gt=0)
    practitioner_id: Optional[int] = Field(None, gt=0)
    anesthesia_type: Optional[str] = Field(None, max_length=100)
    anesthesia_notes: Optional[str] = Field(None, max_length=1000)
    tags: Optional[List[str]] = None

    @model_validator(mode="before")
    @classmethod
    def validate_date_with_status(cls, values):
        """Validate procedure date based on status.

        When status is not provided (partial updates), skip validation to allow
        updating only the date field without requiring status.
        """
        from datetime import timedelta

        if not isinstance(values, dict):
            return values

        date_value = values.get("date")
        status = values.get("status", "").lower() if values.get("status") else ""

        if not date_value:
            return values

        # Convert string date to date object if needed
        if isinstance(date_value, str):
            try:
                date_value = DateType.fromisoformat(date_value)
            except ValueError:
                return values  # Let field validator handle invalid date

        # Skip validation if status is not provided (partial update scenario)
        if not status:
            return values

        # For scheduled or postponed procedures, allow reasonable future dates
        if status in ["scheduled", "postponed"]:
            max_future = DateType.today() + timedelta(days=3650)  # 10 years
            if date_value > max_future:
                raise ValueError("Procedure date cannot be more than 10 years in the future")
            # Allow past dates for scheduled procedures (e.g., rescheduled from past)
            return values

        # For all other statuses (completed, in_progress, cancelled), date should not be in future
        if date_value > DateType.today():
            raise ValueError(f"Procedure date cannot be in the future for {status} procedures")
        return values

    @field_validator("status")
    @classmethod
    def validate_status(cls, v):
        if v is not None:
            valid_statuses = [s.value for s in ProcedureStatus]
            if v.lower() not in valid_statuses:
                raise ValueError(f"Status must be one of: {', '.join(valid_statuses)}")
            return v.lower()
        return v


class ProcedureResponse(ProcedureBase):
    id: int

    class Config:
        from_attributes = True


class ProcedureWithRelations(ProcedureResponse):
    patient: Optional[dict] = None
    practitioner: Optional[dict] = None

    class Config:
        from_attributes = True


class ProcedureSummary(BaseModel):
    id: int
    procedure_name: str
    date: DateType
    status: str
    patient_name: Optional[str] = None
    practitioner_name: Optional[str] = None

    class Config:
        from_attributes = True

File: frontend/src/utils/medicalFormFields/procedure.js
/**
 * Procedure form field configuration
 */

import { tagsFieldConfig } from './shared';

export const procedureFormFields = [
  {
    name: 'procedure_name',
    type: 'text',
    labelKey: 'medical:procedures.procedureName.label',
    placeholderKey: 'medical:procedures.procedureName.placeholder',
    required: true,
    descriptionKey: 'medical:procedures.procedureName.description',
    gridColumn: 8,
  },
  {
    name: 'procedure_type',
    type: 'select',
    labelKey: 'medical:procedures.procedureType.label',
    placeholderKey: 'medical:procedures.procedureType.placeholder',
    descriptionKey: 'medical:procedures.procedureType.description',
    gridColumn: 4,
    clearable: true,
    options: [
      {
        value: 'surgical',
        labelKey: 'medical:procedures.procedureType.options.surgical',
      },
      {
        value: 'diagnostic',
        labelKey: 'medical:procedures.procedureType.options.diagnostic',
      },
      {
        value: 'therapeutic',
        labelKey: 'medical:procedures.procedureType.options.therapeutic',
      },
      {
        value: 'preventive',
        labelKey: 'medical:procedures.procedureType.options.preventive',
      },
      {
        value: 'emergency',
        labelKey: 'medical:procedures.procedureType.options.emergency',
      },
    ],
  },
  {
    name: 'procedure_code',
    type: 'text',
    labelKey: 'medical:procedures.procedureCode.label',
    placeholderKey: 'medical:procedures.procedureCode.placeholder',
    descriptionKey: 'medical:procedures.procedureCode.description',
    gridColumn: 6,
  },
  {
    name: 'procedure_setting',
    type: 'select',
    labelKey: 'medical:procedures.procedureSetting.label',
    placeholderKey: 'medical:procedures.procedureSetting.placeholder',
    descriptionKey: 'medical:procedures.procedureSetting.description',
    gridColumn: 6,
    clearable: true,
    options: [
      {
        value: 'outpatient',
        labelKey: 'medical:procedures.procedureSetting.options.outpatient',
      },
      {
        value: 'inpatient',
        labelKey: 'medical:procedures.procedureSetting.options.inpatient',
      },
      {
        value: 'office',
        labelKey: 'medical:procedures.procedureSetting.options.office',
      },
      {
        value: 'emergency',
        labelKey: 'medical:procedures.procedureSetting.options.emergency',
      },
      {
        value: 'home',
        labelKey: 'medical:procedures.procedureSetting.options.home',
      },
    ],
  },
  {
    name: 'description',
    type: 'textarea',
    labelKey: 'medical:procedures.description.label',
    placeholderKey: 'medical:procedures.description.placeholder',
    descriptionKey: 'medical:procedures.description.description',
    gridColumn: 12,
    minRows: 3,
    maxRows: 5,
  },
  {
    name: 'procedure_date',
    type: 'date',
    labelKey: 'medical:procedures.procedureDate.label',
    placeholderKey: 'medical:procedures.procedureDate.placeholder',
    required: true,
    descriptionKey: 'medical:procedures.procedureDate.description',
    gridColumn: 4,
  },
  {
    name: 'status',
    type: 'select',
    labelKey: 'common:labels.status',
    descriptionKey: 'medical:procedures.status.description',
    gridColumn: 4,
    dynamicOptions: 'statuses',
  },
  {
    name: 'procedure_duration',
    type: 'number',
    labelKey: 'medical:procedures.procedureDuration.label',
    placeholderKey: 'medical:procedures.procedureDuration.placeholder',
    descriptionKey: 'medical:procedures.procedureDuration.description',
    gridColumn: 4,
    min: 1,
  },
  {
    name: 'facility',
    type: 'text',
    labelKey: 'medical:procedures.facility.label',
    placeholderKey: 'medical:procedures.facility.placeholder',
    descriptionKey: 'medical:procedures.facility.description',
    gridColumn: 6,
  },
  {
    name: 'practitioner_id',
    type: 'select',
    labelKey: 'medical:procedures.practitioner.label',
    placeholderKey: 'medical:procedures.practitioner.placeholder',
    descriptionKey: 'medical:procedures.practitioner.description',
    gridColumn: 6,
    clearable: true,
    searchable: true,
    dynamicOptions: 'practitioners',
  },
  {
    name: 'procedure_complications',
    type: 'textarea',
    labelKey: 'medical:procedures.complications.label',
    placeholderKey: 'medical:procedures.complications.placeholder',
    descriptionKey: 'medical:procedures.complications.description',
    gridColumn: 12,
    minRows: 2,
    maxRows: 4,
  },
  {
    name: 'notes',
    type: 'textarea',
    labelKey: 'medical:procedures.notes.label',
    placeholderKey: 'medical:procedures.notes.placeholder',
    descriptionKey: 'medical:procedures.notes.description',
    gridColumn: 12,
    minRows: 3,
    maxRows: 6,
  },
  {
    name: 'anesthesia_type',
    type: 'select',
    labelKey: 'medical:procedures.anesthesiaType.label',
    placeholderKey: 'medical:procedures.anesthesiaType.placeholder',
    descriptionKey: 'medical:procedures.anesthesiaType.description',
    gridColumn: 6,
    clearable: true,
    options: [
      {
        value: 'general',
        labelKey: 'medical:procedures.anesthesiaType.options.general',
      },
      {
        value: 'local',
        labelKey: 'medical:procedures.anesthesiaType.options.local',
      },
      {
        value: 'regional',
        labelKey: 'medical:procedures.anesthesiaType.options.regional',
      },
      {
        value: 'sedation',
        labelKey: 'medical:procedures.anesthesiaType.options.sedation',
      },
      {
        value: 'none',
        labelKey: 'medical:procedures.anesthesiaType.options.none',
      },
    ],
  },
  {
    name: 'anesthesia_notes',
    type: 'text',
    labelKey: 'medical:procedures.anesthesiaNotes.label',
    placeholderKey: 'medical:procedures.anesthesiaNotes.placeholder',
    descriptionKey: 'medical:procedures.anesthesiaNotes.description',
    gridColumn: 6,
  },
  tagsFieldConfig,
];
